core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -0,0 +1,247 @@
+   */
+  public void mastersStartedCallback() {
+    if (mExclusiveOnlyDeadlineMs == -1) {
+      mExclusiveOnlyDeadlineMs = System.currentTimeMillis()
+          + ServerConfiguration.getMs(PropertyKey.MASTER_BACKUP_STATE_LOCK_EXCLUSIVE_DURATION);
+    }
+  }
+
+  /**
+   * Locks the state shared.
+   *
+   * Calling thread might be interrupted by this manager,
+   * if it found to be waiting for the shared lock under when:
+   *  - backup is exiting grace-cycle and entering the lock permanently
+   *  - backup is in progress
+   *
+   * @return the lock resource
+   * @throws InterruptedException
+   */
+  public LockResource lockShared() throws InterruptedException {
+    // Do not allow taking shared lock during safe-mode.
+    if (System.currentTimeMillis() < mExclusiveOnlyDeadlineMs) {
+      String safeModeMsg = String.format(
+          "Master still in exclusive-only phase for the state-lock."
+              + "Please see documentation for %s.",
+          PropertyKey.Name.MASTER_BACKUP_STATE_LOCK_EXCLUSIVE_DURATION);
+      throw new IllegalStateException(safeModeMsg);
+    }
+    // Register thread for interrupt cycle.
+    mSharedWaitersAndHolders.add(Thread.currentThread());
+    // Grab the lock interruptibly.
+    mStateLock.readLock().lockInterruptibly();
+    // Return the resource.
+    // Register an action to remove the thread from holders registry before releasing the lock.
+    return new LockResource(mStateLock.readLock(), false, false, () -> {
+      mSharedWaitersAndHolders.remove(Thread.currentThread());
+    });
+  }
+
+  /**
+   * Locks the state exclusively.
+   *
+   * @param lockOptions exclusive lock options
+   * @return the lock resource
+   * @throws TimeoutException if locking times out
+   * @throws InterruptedException if interrupting during locking
+   */
+  public LockResource lockExclusive(StateLockOptions lockOptions)
+      throws TimeoutException, InterruptedException {
+    // Run the grace cycle.
+    StateLockOptions.GraceMode graceMode = lockOptions.getGraceMode();
+    boolean lockAcquired = false;
+    long deadlineMs = System.currentTimeMillis() + lockOptions.getGraceCycleTimeoutMs();
+    while (System.currentTimeMillis() < deadlineMs) {
+      if (mStateLock.writeLock().tryLock(lockOptions.getGraceCycleTryMs(), TimeUnit.MILLISECONDS)) {
+        lockAcquired = true;
+        break;
+      } else {
+        long remainingWaitMs = deadlineMs - System.currentTimeMillis();
+        if (remainingWaitMs > 0) {
+          Thread.sleep(Math.min(lockOptions.getGraceCycleSleepMs(), remainingWaitMs));
+        }
+      }
+    }
+    if (lockAcquired) { // Lock was acquired within grace-cycle.
+      activateInterruptCycle();
+    } else { // Lock couldn't be acquired by grace-cycle.
+      if (graceMode == StateLockOptions.GraceMode.TIMEOUT) {
+        throw new TimeoutException(
+            ExceptionMessage.STATE_LOCK_TIMED_OUT.getMessage(lockOptions.getGraceCycleTimeoutMs()));
```

Comment:
Will it be easy to search for log message, where will this be logged?

Reply:
This will end up being caught at `RpcUtils.callAndReturn`, which prints an error-log for runtime exceptions.